### DISCUSSION TITLE
List comprehensions

### DISCUSSION THREAD

**Discussion by ntrel** [Neat-Lang/neat#11] [General]
There's an `argmax` list comprehension on the main page:
```d
    auto text = readText(args[1]);
    string longestLine = [
        argmax(line.strip.length) line
        for line in text.split("\n")];
```
I find it hard to understand. Shouldn't a list comprehension conceptually generate an array rather than a single value? Normally `arr.argmax` should return the index of the element in `arr` with the largest value? So why does `argmax` take a line length as its argument not an array? And why is there a trailing `line` after that? I find it clearer to write:
```d
auto lines = text.split("\n");
auto longestLine = lines[[line.length for line in lines].argmax];
```
Perhaps the named algorithms could support trailing algorithm syntax instead:
```d
[a > 2 for a in arr].any
[a > 1 for a in arr].all
[a in arr where a > 2].count
```

Also a list comprehension could produce a range and the algorithm suffix could just be a range function rather than built-in. Then any custom function could be used.

**Comment by ntrel** [comment]
> I find it clearer to write:
> 
> ```d
> auto lines = text.split("\n");
> auto longestLine = lines[[line.length for line in lines].argmax];
> ```

Actually if we had `maxElement` with a predicate I would write it like this:
```d
string longestLine = text.split("\n").maxElement(line => line.strip.length);
// docs:
string longestLine = [argmax(line.strip.length) line for line in text.split("\n")];
```

**Reply by FeepingCreature** [reply]
The list comprehension macro is my equivalent of C#'s LINQ. I wrote it before I added lambdas, so it does anything and everything. I agree that the `std.algorithm` form would be clearer, but I feel people already know that from D: what I want to get across with using the list comprehension example is "You, yes you, can add features that are just as powerful as this."

**Reply by FeepingCreature** [reply]
That said, I fully agree that `maxElement` should exist.
